vcd/resource_vcd_nsxt_edgegateway_dns.go

```diff
@@ -0,0 +1,380 @@
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+	},
+}
+
+func resourceVcdNsxtEdgegatewayDnsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return resourceVcdNsxtEdgegatewayDnsCreateUpdate(ctx, d, meta, "create")
+}
+
+func resourceVcdNsxtEdgegatewayDnsUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return resourceVcdNsxtEdgegatewayDnsCreateUpdate(ctx, d, meta, "update")
+}
+
+func resourceVcdNsxtEdgegatewayDnsCreateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}, origin string) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	unlock, err := vcdClient.lockParentVdcGroupOrEdgeGateway(d)
+	if err != nil {
+		return diag.Errorf("[edge gateway dns %s] %s", origin, err)
+	}
+	defer unlock()
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if err != nil {
+		return diag.Errorf("[edge gateway dns %s] error retrieving Edge Gateway: %s", origin, err)
+	}
+
+	dns, err := nsxtEdge.GetDnsConfig()
+	if err != nil {
+		return diag.Errorf("[edge gateway dns %s] error getting current DNS configuration: %s", origin, err)
+	}
+
+	dnsConfig, err := getNsxtEdgeGatewayDnsConfig(d, vcdClient)
+	if err != nil {
+		return diag.Errorf("[edge gateway dns %s] error getting DNS configuration from schema: %s", origin, err)
+	}
+
+	updatedDns, err := dns.Update(dnsConfig)
+	if err != nil {
+		return diag.Errorf("[edge gateway dns %s] error updating DNS configuration: %s", origin, err)
+	}
+
+	d.SetId(updatedDns.EdgeGatewayId)
+
+	return resourceVcdNsxtEdgegatewayDnsRead(ctx, d, meta)
+}
+
+func resourceVcdNsxtEdgegatewayDnsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return genericVcdNsxtEdgegatewayDnsRead(ctx, d, meta, "resource")
+}
+
+func genericVcdNsxtEdgegatewayDnsRead(_ context.Context, d *schema.ResourceData, meta interface{}, origin string) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	orgName := d.Get("org").(string)
+
+	var nsxtEdge *govcd.NsxtEdgeGateway
+	var err error
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+	if d.Id() == "" && edgeGatewayId == "" {
+		return diag.Errorf("id wasn't provided for Edge Gateway DNS")
+	}
+	if edgeGatewayId == "" {
+		nsxtEdge, err = vcdClient.GetNsxtEdgeGatewayById(orgName, d.Id())
+	} else {
+		nsxtEdge, err = vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	}
+	if err != nil {
+		if origin == "resource" && govcd.ContainsNotFound(err) {
+			// When parent Edge Gateway is not found - this resource is also not found and should be
+			// removed from state
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	dns, err := nsxtEdge.GetDnsConfig()
+	if err != nil {
+		return diag.Errorf("[edge gateway dns read] error retrieving NSX-T Edge Gateway DNS config: %s", err)
+	}
```

Comment:
One more (and probably last) thing about the "read" part from me. If I create the resource and remove that configuration (click the `delete` button in UI), then doing a `terraform apply` returns an error as per below. 

We could detect when `dns, err := nsxtEdge.GetDnsConfig()` is empty and do `d.SetId("")` if it is a resource - user would get a new plan and it would simply recreate it.

```
 # vcd_nsxt_edgegateway_dns.TestAccVcdNsxtEdgegatewayDnsIpSpaces will be updated in-place
  ~ resource "vcd_nsxt_edgegateway_dns" "TestAccVcdNsxtEdgegatewayDnsIpSpaces" {
      ~ enabled              = false -> true
        id                   = "urn:vcloud:gateway:f0e97d19-67a3-4c5c-85c9-6af9c61034f5"
      + snat_rule_ip_address = "11.11.11.100"
        # (2 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
vcd_nsxt_edgegateway_dns.TestAccVcdNsxtEdgegatewayDnsIpSpaces: Modifying... [id=urn:vcloud:gateway:f0e97d19-67a3-4c5c-85c9-6af9c61034f5]
╷
│ Error: [edge gateway dns update] error updating DNS configuration: error waiting completion of task (https://_HOST_/api/task/685c3f14-b6a9-4fae-b81a-8d722b48600f): task did not complete successfully:  [400:BAD_REQUEST] - [ 30a97b53-cadd-49a7-aa95-6ab25e539987 ] DNS forwarder zone d5e141c6-abf5-4b1f-908a-b36f343eea96 does not exist.
│ 
│   with vcd_nsxt_edgegateway_dns.TestAccVcdNsxtEdgegatewayDnsIpSpaces,
│   on vcd.TestAccVcdNsxtEdgegatewayDnsIpSpacesstep2.tf line 107, in resource "vcd_nsxt_edgegateway_dns" "TestAccVcdNsxtEdgegatewayDnsIpSpaces":
│  107: resource "vcd_nsxt_edgegateway_dns" "TestAccVcdNsxtEdgegatewayDnsIpSpaces" {
│ 
```

Reply:
Added a check
```go
dns, err := nsxtEdge.GetDnsConfig()
if err != nil {
	return diag.Errorf("[edge gateway dns %s] error getting current DNS configuration: %s", origin, err)
}

if dns.DefaultForwarderZone == nil {
  d.SetId("")
  return nil
}
```
to the Read function, DefaultForwarderZone must exist if the DNS exists, so this should be a valid check.